### ISSUE TITLE
SimpleResponseParser fails when directives reference absent booleans with default values

### ISSUE THREAD

**Issue by DSteve595** [apollographql/apollo-android#2703]
**Version**
2.4.1

**Description**
Given a query that has a boolean arg with a default value, e.g.
```graphql
query MyQuery($myBool: Boolean = true) {
    getCityByName(name: "London") {
        id @include(if: $myBool)
    }
}
```

Parsing this query via `ApolloClient` works as expected. However, using `Operation::parse` fails with this exception:
```
kotlin.TypeCastException: null cannot be cast to non-null type kotlin.Boolean
	at com.apollographql.apollo.api.internal.SimpleResponseReader.shouldSkip(SimpleResponseReader.kt:133)
	at com.apollographql.apollo.api.internal.SimpleResponseReader.readCustomType(SimpleResponseReader.kt:99)
	at MyQuery$GetCityByName$Companion.invoke(MyQuery.kt:123)
	at MyQuery$Data$Companion$invoke$1$getCityByName$1.invoke(MyQuery.kt:153)
	at MyQuery$Data$Companion$invoke$1$getCityByName$1.invoke(MyQuery.kt:145)
	at com.apollographql.apollo.api.internal.ResponseReader$readObject$1.read(ResponseReader.kt:26)
	at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:80)
	at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readObject(ResponseReader.kt:24)
	at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:11)
	at MyQuery$Data$Companion.invoke(MyQuery.kt:152)
	at MyQuery$responseFieldMapper$$inlined$invoke$1.map(ResponseFieldMapper.kt:22)
	at com.apollographql.apollo.api.internal.SimpleOperationResponseParser.parse(SimpleOperationResponseParser.kt:39)
	at com.apollographql.apollo.api.internal.SimpleOperationResponseParser.parse(SimpleOperationResponseParser.kt:26)
	at MyQuery.parse(MyQuery.kt:66)
	at MyQuery.parse(MyQuery.kt:73)
```

### [Sample project](https://github.com/apollographql/apollo-android/files/5443531/apollo-directive-optional-bug.zip)
Run `./gradlew run` to trigger the issue.

**Comment by martinbonnin** [comment]
Thanks for filing this. It looks related to https://github.com/apollographql/apollo-android/issues/2686

**Comment by martinbonnin** [comment]
See pull request in #2704 for a fix. 


Side note: for this particular exemple, the [spec recommends making the Boolean as non-nullable to avoid ambiguity](http://spec.graphql.org/draft/#sec-All-Variable-Usages-are-Allowed.Allowing-optional-variables-when-default-values-exist).
It would be nice to output a warning during validation in these cases.


```graphql
# myBool should be non-optional here
query MyQuery($myBool: Boolean! = true) {
    getCityByName(name: "London") {
        id @include(if: $myBool)
    }
}
```

**Comment by DSteve595** [comment]
Making the boolean non-nullable is what made sense to me, so it's actually what I tried first. But (at least on GraphiQL), I get this:
![image](https://user-images.githubusercontent.com/1078569/97337791-f35d9400-183d-11eb-9a6a-a0b92bf2b50b.png)

**Comment by martinbonnin** [comment]
Interesting. Playground does the same indeed. Looking at [this other issue](https://github.com/rmosolgo/graphql-ruby/issues/2356) it looks like this is more of a tooling issue. Does the query run if you hit play despite the red warning?

**Comment by DSteve595** [comment]
Nope, same error:
```json
{
  "message": "Query validation error: Invalid query: ['Variable \"$includeSubredditInPosts\" of type \"Boolean!\" is required and will not use the default value. Perhaps you meant to use type \"Boolean\".']"
}

**Comment by martinbonnin** [comment]
Interesting. Optional it is then! And certainly not worth adding a warning in the compiler if fixing the warning breaks the query 😬

**Comment by DSteve595** [comment]
Hmm.. I can confirm that 2.4.2 fixes this in my sample project, but I'm still seeing the error in a bigger project. Here's the full stack trace. If more info is required (like the full query), would prefer to message privately.

```
kotlin.TypeCastException: null cannot be cast to non-null type kotlin.Boolean
    at com.apollographql.apollo.api.internal.SimpleResponseReader.shouldSkip(SimpleResponseReader.kt:133)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:72)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readObject(ResponseReader.kt:24)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:11)
    at com.reddit.fragment.PostContentFragment$AsSubredditPost$Companion.invoke(PostContentFragment.kt:2583)
    at com.reddit.fragment.PostContentFragment$Companion$invoke$1$asSubredditPost$1.invoke(PostContentFragment.kt:525)
    at com.reddit.fragment.PostContentFragment$Companion$invoke$1$asSubredditPost$1.invoke(PostContentFragment.kt:275)
    at com.apollographql.apollo.api.internal.ResponseReader$readFragment$1.read(ResponseReader.kt:36)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:125)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readFragment(ResponseReader.kt:34)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:11)
    at com.reddit.fragment.PostContentFragment$Companion.invoke(PostContentFragment.kt:524)
    at com.reddit.fragment.PostFragment$Fragments$Companion$invoke$1$postContentFragment$1.invoke(PostFragment.kt:91)
    at com.reddit.fragment.PostFragment$Fragments$Companion$invoke$1$postContentFragment$1.invoke(PostFragment.kt:84)
    at com.apollographql.apollo.api.internal.ResponseReader$readFragment$1.read(ResponseReader.kt:36)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:125)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readFragment(ResponseReader.kt:34)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:11)
    at com.reddit.fragment.PostFragment$Fragments$Companion.invoke(PostFragment.kt:90)
    at com.reddit.fragment.PostFragment$Companion.invoke(PostFragment.kt:65)
    at com.reddit.queries.PopularPostsQuery$Node$Fragments$Companion$invoke$1$postFragment$1.invoke(PopularPostsQuery.kt:277)
    at com.reddit.queries.PopularPostsQuery$Node$Fragments$Companion$invoke$1$postFragment$1.invoke(PopularPostsQuery.kt:270)
    at com.apollographql.apollo.api.internal.ResponseReader$readFragment$1.read(ResponseReader.kt:36)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:125)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readFragment(ResponseReader.kt:34)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readFragment(SimpleResponseReader.kt:11)
    at com.reddit.queries.PopularPostsQuery$Node$Fragments$Companion.invoke(PopularPostsQuery.kt:276)
    at com.reddit.queries.PopularPostsQuery$Node$Companion.invoke(PopularPostsQuery.kt:252)
    at com.reddit.queries.PopularPostsQuery$Edge$Companion$invoke$1$node$1.invoke(PopularPostsQuery.kt:311)
    at com.reddit.queries.PopularPostsQuery$Edge$Companion$invoke$1$node$1.invoke(PopularPostsQuery.kt:302)
    at com.apollographql.apollo.api.internal.ResponseReader$readObject$1.read(ResponseReader.kt:26)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:80)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readObject(ResponseReader.kt:24)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:11)
    at com.reddit.queries.PopularPostsQuery$Edge$Companion.invoke(PopularPostsQuery.kt:310)
    at com.reddit.queries.PopularPostsQuery$Posts$Companion$invoke$1$edges$1$1.invoke(PopularPostsQuery.kt:359)
    at com.reddit.queries.PopularPostsQuery$Posts$Companion$invoke$1$edges$1$1.invoke(PopularPostsQuery.kt:343)
    at com.apollographql.apollo.api.internal.ResponseReader$ListItemReader$readObject$1.read(ResponseReader.kt:80)
    at com.apollographql.apollo.api.internal.SimpleResponseReader$ListItemReader.readObject(SimpleResponseReader.kt:191)
    at com.apollographql.apollo.api.internal.ResponseReader$ListItemReader$DefaultImpls.readObject(ResponseReader.kt:78)
    at com.apollographql.apollo.api.internal.SimpleResponseReader$ListItemReader.readObject(SimpleResponseReader.kt:158)
    at com.reddit.queries.PopularPostsQuery$Posts$Companion$invoke$1$edges$1.invoke(PopularPostsQuery.kt:358)
    at com.reddit.queries.PopularPostsQuery$Posts$Companion$invoke$1$edges$1.invoke(PopularPostsQuery.kt:343)
    at com.apollographql.apollo.api.internal.ResponseReader$readList$1.read(ResponseReader.kt:46)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readList(SimpleResponseReader.kt:94)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readList(ResponseReader.kt:44)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readList(SimpleResponseReader.kt:11)
    at com.reddit.queries.PopularPostsQuery$Posts$Companion.invoke(PopularPostsQuery.kt:357)
    at com.reddit.queries.PopularPostsQuery$PostFeed$Companion$invoke$1$posts$1.invoke(PopularPostsQuery.kt:423)
    at com.reddit.queries.PopularPostsQuery$PostFeed$Companion$invoke$1$posts$1.invoke(PopularPostsQuery.kt:390)
    at com.apollographql.apollo.api.internal.ResponseReader$readObject$1.read(ResponseReader.kt:26)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:80)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readObject(ResponseReader.kt:24)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:11)
    at com.reddit.queries.PopularPostsQuery$PostFeed$Companion.invoke(PopularPostsQuery.kt:422)
    at com.reddit.queries.PopularPostsQuery$Data$Companion$invoke$1$postFeed$1.invoke(PopularPostsQuery.kt:459)
    at com.reddit.queries.PopularPostsQuery$Data$Companion$invoke$1$postFeed$1.invoke(PopularPostsQuery.kt:452)
    at com.apollographql.apollo.api.internal.ResponseReader$readObject$1.read(ResponseReader.kt:26)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:80)
    at com.apollographql.apollo.api.internal.ResponseReader$DefaultImpls.readObject(ResponseReader.kt:24)
    at com.apollographql.apollo.api.internal.SimpleResponseReader.readObject(SimpleResponseReader.kt:11)
    at com.reddit.queries.PopularPostsQuery$Data$Companion.invoke(PopularPostsQuery.kt:458)
    at com.reddit.queries.PopularPostsQuery$responseFieldMapper$$inlined$invoke$1.map(ResponseFieldMapper.kt:22)
    at com.apollographql.apollo.api.internal.SimpleOperationResponseParser.parse(SimpleOperationResponseParser.kt:39)
    at com.apollographql.apollo.api.internal.SimpleOperationResponseParser.parse(SimpleOperationResponseParser.kt:26)
    at com.reddit.queries.PopularPostsQuery.parse(PopularPostsQuery.kt:139)
    at com.reddit.queries.PopularPostsQuery.parse(PopularPostsQuery.kt:146)
```

**Comment by DSteve595** [comment]
It seems to happen when passing `Input.absent()` as the arg. Since `Input` already has `defined: Boolean`, maybe the GraphQL-provided default arg should be used when `defined == false`?

**Comment by martinbonnin** [comment]
Fixed with #2741
